src/backend/cpu/kernel/fft.hpp

```diff
@@ -121,7 +121,7 @@ void fft_r2c(Param<Tc> out, const af::dim4 oDataDims, CParam<Tr> in,
     plan = transform.create(rank, t_dims, (int)batch, (Tr *)in.get(), in_embed,
                             (int)istrides[0], (int)istrides[rank],
                             (ctype_t *)out.get(), out_embed, (int)ostrides[0],
-                            (int)ostrides[rank], FFTW_ESTIMATE);
+                            (int)ostrides[rank], FFTW_ESTIMATE | FFTW_PRESERVE_INPUT);
```

Comment:
Is this flag alone enough to fix the problem? Why copy the array yet again?

Reply:
This only works in real to complex ffts. The complex 2 real function will return a NULL plan because FFTW does not have an algorithm for this.

> FFTW_PRESERVE_INPUT specifies that an out-of-place transform must not change its input array. This is ordinarily the default, except for c2r and hc2r (i.e. complex-to-real) transforms for which FFTW_DESTROY_INPUT is the default. In the latter cases, passing FFTW_PRESERVE_INPUT will attempt to use algorithms that do not destroy the input, at the expense of worse performance; for multi-dimensional c2r transforms, however, no input-preserving algorithms are implemented and the planner will return NULL if one is requested.

Reply:
Yikes! I'm so glad ArrayFire has separate functions for in-place transforms and it is clear from the name of the function of what to expect.